### ISSUE TITLE
A variable explorer for notebook? (like MATLAB's Workspace)

### ISSUE THREAD

**Issue by cqcn1991** [ipython-contrib/IPython-notebook-extensions#623]
I want to know if there is anything like MATLAB's Workspace? (or Spyder's Variable Explorer). So we can easily copy the values out. Something look like this:

[![enter image description here][1]][1]

  [1]: http://i.stack.imgur.com/XyTKP.png

-----

I ask it here (https://github.com/jupyter/notebook/issues/1516),  Carreau gives a very good example. In short, you can get the variables by this:

```
variables = %who_ls
for v in variables:
    print v, type(eval(v)).__name__, str(eval(v))
```

Another thing worth considering is `spyderlib.plugins.variableexplorer` from Spyder. This could be seen at http://stackoverflow.com/questions/2960864/how-can-i-save-all-the-variables-in-the-current-python-session

**Comment by janschulz** [comment]
rodeo has such a thing: 
https://github.com/yhat/rodeo/blob/921ad700c432f5510f1b368c16d1acc7b361de2a/src/browser/jsx/containers/variable-viewer.jsx
https://github.com/yhat/rodeo/blob/0b4a9ed5acf052fa9e2eca9427263b5dc44c187d/src/node/kernels/python/langs/python-patch.py

I would also love to see such a thing! On the other hand: isn't that what the jupyterhub stuff is also about?

**Comment by cqcn1991** [comment]
@janschulz Never thought of Rodeo, I thought it was about desktop version. 
Never heard of jupyterlab, is it this? https://github.com/jupyter/jupyterlab

**Comment by janschulz** [comment]
rodeo isn't a notebook app, but also uses the kernel infrastructure and has a browser UI (Rodeo (and RStudio on which rodeo is based) are basically a browser window).

**Comment by janschulz** [comment]
The original answer in the notebook repo linked to https://github.com/ipython/ipywidgets/blob/master/docs/source/examples/Variable%20Inspector.ipynb -> doesn't look so complicated :-)

**Comment by cqcn1991** [comment]
@janschulz Yes, that widget is pretty amazing and simple. The problem is that how to make it "copy and pastable", i.e. have table display UI for array and matrix. 

Addtionally, I personally want a Ruby on Rails-like framework for Jupyter Notebook, in which have pre-load common use nbextensions, libraries and good file organizations. It can help people get to know useful features, extentions at the beginning, saving people tons of time for configuration.

**Comment by flying-sheep** [comment]
to be cross-language-compatible we need a protocol extension AFAICS:

1. a request to the kernel to return the names and types of variables.
2. a request to the kernel to display or open an interactive editor for a variable (i.e. the kernel can reply with a open-pager message or create a widget able to edit the variable)

**Comment by flying-sheep** [comment]
@cqcn1991 to get what you want right now, you can do something like

```py
from IPython.core.page import page
from IPython.core.formatters import format_display_data

page(format_display_data(thing))
``

**Comment by cqcn1991** [comment]
@flying-sheep what is `thing` in `page(format_display_data(thing, ['text/plain', 'text/html']))`?

**Comment by flying-sheep** [comment]
the variable you want to display. to get it there’s a bit more work involved:

1. you have to set up JS event listeners for the cells that send a comm request to the kernel
2. you have to set up event listeners in python that react to that comm request and do `thing = ipython.kernel.shell.user_ns[varname]`

**Comment by PatrickChrist** [comment]
Would be a great feature!

**Comment by jsekamane** [comment]
How about the [Variable Inspector](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/varInspector/README.html) extension?

**Comment by ashimb9** [comment]
Does the variable inspector work in Jupyterlab?

**Comment by jcb91** [comment]
> Does the variable inspector work in Jupyterlab?
No, jupyterlab uses a different (and so far still constantly-changing) javascript model, so I know of no extensions that work there yet

**Comment by jfbercher** [comment]
A varInspector is planned in the future in `JupyterLab`, see the [roadmap](https://github.com/jupyter/roadmap/blob/master/jupyterlab.md). 

Also, it is planned in the notebook development  [roadmap for Jupyter v 5+](https://github.com/jupyter/roadmap/blob/master/notebook.md) to port/bridge popular nbextension "*we will continue to work on porting nbextensions to JupyterLab and building the bridge layers.*" and "*At least some important, popular extensions are ported / bridged to Jupyter Lab*".  Of course we will be happy to port or help, but for now the interface seems not quite stable and/or I don't find easy keys.

**Comment by biocyberman** [comment]
> How about the [Variable Inspector](http://jupyter-contrib-nbextensions.readthedocs.io/en/latest/nbextensions/varInspector/README.html) extension?

This works for simple variables, but a better presentation (i.e. like that of in Spyder) of dataframes/matrices/vectors is much appreciated. 

Alternatively, I also tried running jupyter notebooks inside spyder, this did not work well.
